packages/websocket/src/index.ts

```diff
@@ -1,72 +1,199 @@
-import { createSignal, onCleanup } from "solid-js";
+import { createMemo, createSignal, onCleanup, type Accessor } from "solid-js";
 
-export type WebsocketState = 0 | 1 | 2 | 3;
+export type WSMessage = string | ArrayBufferLike | ArrayBufferView | Blob;
 
 /**
- * Handles opening managing and reconnecting to a Websocket.
- *
- * @param url - Path to the websocket server
- * @param onData - A function supplied that messages will be reported to
- * @param onError - A function supplied that errors will be reported to
- * @param procotols - List of protocols to support
- * @param reconnectLimit - Amount of reconnection attempts
- * @param reconnectInterval - Time in between connection attempts
- * @return Returns an array containing websocket management options
+ * opens a web socket connection with a queued send
+ * ```ts
+ * const ws = makeWS("ws:localhost:5000");
+ * createEffect(() => ws.send(serverMessage()));
+ * onCleanup(() => ws.close());
+ * ```
+ * Will not throw if you attempt to send messages before the connection opened; instead, it will enqueue the message to be sent when the connection opens.
  *
- * @example
+ * It will not close the connection on cleanup. To do that, use `createWS`.
+ */
+export const makeWS = (
+  url: string,
+  protocols?: string | string[],
+  sendQueue: WSMessage[] = [],
+): WebSocket => {
+  const ws: WebSocket = new WebSocket(url, protocols);
+  const _send = ws.send.bind(ws);
+  ws.send = (msg: WSMessage) =>
+    ws.readyState == 1 ? _send(msg) : sendQueue.push(msg);
+  ws.addEventListener('open', () => {
+    while (sendQueue.length) _send(sendQueue.shift()!);
+  });
+  return ws;
+}
+
+/**
+ * Extends a WebSocket with a message signal:
  * ```ts
- * const [ connect, disconnect ] = createWebsocket('http://localhost', '', 3, 5000);
+ * const ws = createMessageWS(makeWS("ws://localhost:5000"));
+ * createEffect(on(ws.message, (msg) => console.log(msg)), { defer: true });
  * ```
  */
-const createWebsocket = (
+export const createMessageWS = (
+  ws: WebSocket
+): WebSocket & { message: Accessor<WSMessage | undefined> } => {
```

Comment:
Cannot it just return the message instead of mutating the input? This seems like an odd pattern considering that the type won't be reflected in the original object.
```ts
const ws = makeWS("ws://localhost:5000")
createMessageWS(ws)
ws.message() // works, but ts doesn't know about it
```

It also doesn't support setting initial value, which could be useful.

Also, we have `createEventSignal` in `@solid-primitives/event-listener`. Why not promote using that as it can be used with any event target?

```ts
import { createEventSignal } from "@solid-primitives/event-listener"

const ws = makeWS("ws://localhost:5000")
const message = createEventSignal(ws, "message")

message() // WSMessage | undefined
```

Reply:
I removed this one and settled on documenting the use of `createEventSignal`.